### ISSUE TITLE
Confusing recommendation to use sync_dist=True even with TorchMetrics

### ISSUE THREAD

**Issue by srprca** [Lightning-AI/pytorch-lightning#20153]
### Bug description

Hello!

When I train and validate a model in a multi-GPU setting (HPC, sbatch job that requests multiple GPUs on a single node), I use `self.log(..., sync_dist=True)` when logging PyTorch losses, and don't specify any value for `sync_dist` when logging metrics from TorchMetrics library. However, I still get warnings like
```
...
It is recommended to use `self.log('val_mean_recall', ..., sync_dist=True)` when logging on epoch level in distributed setting to accumulate the metric across devices.
...
It is recommended to use `self.log('val_bg_recall', ..., sync_dist=True)` when logging on epoch level in distributed setting to accumulate the metric across devices.
```
These specific messages correspond to logging `tmc.MulticlassRecall(len(self.task.class_names), average="macro", ignore_index=self.metric_ignore_index)` and individual components of `tmc.MulticlassRecall(len(self.task.class_names), average="none", ignore_index=self.metric_ignore_index)`.

Full code listing for metric object definitions and logging is provided in the "reproducing the bug" section.

As I understand from a note [here](https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html#logging-torchmetrics), and from discussion [here](https://github.com/Lightning-AI/pytorch-lightning/discussions/6501#discussioncomment-553152), one doesn't typically need to explicitly use `sync_dist` when using TorchMetrics.

I wonder if I still need to enable `sync_dist=True` as advised in the warnings due to some special case that I am not aware about, or should I follow the docs and keep it as is? In any case, this is probably a bug, either in documentation, or in warning code.

Thank you!

### What version are you seeing the problem on?

2.3.0

### How to reproduce the bug

```python
self.val_metric_funs = tm.MetricCollection(
                {
                    "cm_normalize_all": tmc.MulticlassConfusionMatrix(
                        len(self.task.class_names),
                        ignore_index=self.metric_ignore_index,
                        normalize="all",
                    ),
                    "recall_average_macro": tmc.MulticlassRecall(
                        len(self.task.class_names),
                        average="macro",
                        ignore_index=self.metric_ignore_index,
                    ),
                    "recall_average_none": tmc.MulticlassRecall(
                        len(self.task.class_names),
                        average="none",
                        ignore_index=self.metric_ignore_index,
                    ),
                    "precision_average_macro": tmc.MulticlassPrecision(
                        len(self.task.class_names),
                        average="macro",
                        ignore_index=self.metric_ignore_index,
                    ),
                    "precision_average_none": tmc.MulticlassPrecision(
                        len(self.task.class_names),
                        average="none",
                        ignore_index=self.metric_ignore_index,
                    ),
                    "f1_average_macro": tmc.MulticlassF1Score(
                        len(self.task.class_names),
                        average="macro",
                        ignore_index=self.metric_ignore_index,
                    ),
                    "f1_average_none": tmc.MulticlassF1Score(
                        len(self.task.class_names),
                        average="none",
                        ignore_index=self.metric_ignore_index,
                    ),
                }
            )
```

```
if not sanity_check:
            for metric_name, metric in metrics.items():
                metric_fun = self.val_metric_funs[metric_name]
                metric_name_ = metric_name.split("_")[0]
                if isinstance(metric_fun, tmc.MulticlassConfusionMatrix):
                    for true_class_num in range(metric.shape[0]):
                        true_class_name = self.task.class_names[true_class_num]
                        for pred_class_num in range(metric.shape[1]):
                            pred_class_name = self.task.class_names[pred_class_num]
                            self.log(
                                f"val_true_{true_class_name}_pred_{pred_class_name}_cm",
                                metric[true_class_num, pred_class_num].item(),
                                on_step=False,
                                on_epoch=True,
                                logger=True,
                            )
                elif isinstance(
                    metric_fun,
                    (
                        tmc.MulticlassRecall,
                        tmc.MulticlassPrecision,
                        tmc.MulticlassF1Score,
                    ),
                ):
                    if metric_fun.average == "macro":
                        self.log(
                            f"val_mean_{metric_name_}",
                            metric.item(),
                            on_step=False,
                            on_epoch=True,
                            logger=True,
                        )
                    elif metric_fun.average == "none":
                        for class_num, metric_ in enumerate(metric):
                            class_name = self.task.class_names[class_num]
                            self.log(
                                f"val_{class_name}_{metric_name_}",
                                metric_.item(),
                                on_step=False,
                                on_epoch=True,
                                logger=True,
                            )
                    else:
                        raise NotImplementedError(
                            f"Code for logging metric {metric_name} is not implemented"
                        )
                else:
                    raise NotImplementedError(
                        f"Code for logging metric {metric_name} is not implemented"
                    )
```

### Error messages and logs

```
...
It is recommended to use `self.log('val_mean_recall', ..., sync_dist=True)` when logging on epoch level in distributed setting to accumulate the metric across devices.
...
It is recommended to use `self.log('val_bg_recall', ..., sync_dist=True)` when logging on epoch level in distributed setting to accumulate the metric across devices.
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version: 2.3.0
#- PyTorch Version: 2.3.1
#- Python version: 3.11.9
#- OS: Linux
#- CUDA/cuDNN version: 11.8
#- How you installed Lightning: conda-forge
```

</details>

cc @carmocca

**Comment by awaelchli** [comment]
Yes that's right, the warning shouldn't occur when logging TorchMetrics. Does it occur only with MetricCollection or a regular too Metric?

**Comment by srprca** [comment]
Thank you for your reply!

I will be able to check this tomorrow, and will report back.

Meanwhile, my second suspicion is that since I log `metric.item()` objects, is it possible that somehow `self.log` doesn't recognize these objects as originating from TorchMetrics, and sees them as "generic" numbers or tensors...?

I will try to check this hypothesis, too.

**Comment by awaelchli** [comment]
If you pass in scalar tensors then not of course. Then the warning is normal and expected. For logging TorchMetrics you would just pass in the metric directly into `self.log`. You can find a guide here: https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html

**Comment by srprca** [comment]
Actually, this still happens when I remove all the metrics that I logged using `.item()`: now, with the following metrics definition:

```
self.val_metric_funs = tm.MetricCollection(
                {
                    "recall_average_macro": tmc.MulticlassRecall(
                        len(self.task.class_names),
                        average="macro",
                        ignore_index=self.metric_ignore_index,
                    ),
                    "precision_average_macro": tmc.MulticlassPrecision(
                        len(self.task.class_names),
                        average="macro",
                        ignore_index=self.metric_ignore_index,
                    ),
                    "f1_average_macro": tmc.MulticlassF1Score(
                        len(self.task.class_names),
                        average="macro",
                        ignore_index=self.metric_ignore_index,
                    ),
                }
            )
```

and the following logging code:

```
metrics = self.val_metric_funs(logits, annotator_masks_mixed)
        if not sanity_check:
            for metric_name, metric in metrics.items():
                metric_fun = self.val_metric_funs[metric_name]
                metric_name_ = metric_name.split("_")[0]
                if isinstance(
                    metric_fun,
                    (
                        tmc.MulticlassRecall,
                        tmc.MulticlassPrecision,
                        tmc.MulticlassF1Score,
                    ),
                ):
                    if metric_fun.average == "macro":
                        self.log(
                            f"val_mean_{metric_name_}",
                            metric,
                            on_step=False,
                            on_epoch=True,
                            logger=True,
                        )
                    else:
                        raise NotImplementedError(
                            f"Code for logging metric {metric_name} is not implemented"
                        )
                else:
                    raise NotImplementedError(
                        f"Code for logging metric {metric_name} is not implemented"
                    )
```

I get

```
.../miniforge3/envs/main/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/logger_connector/result.py:439: It is recommended to use `self.log('val_mean_f1', ..., sync_dist=True)` when logging on epoch level in distributed setting to accumulate the metric across devices.
.../miniforge3/envs/main/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/logger_connector/result.py:439: It is recommended to use `self.log('val_mean_precision', ..., sync_dist=True)` when logging on epoch level in distributed setting to accumulate the metric across devices.
.../miniforge3/envs/main/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/logger_connector/result.py:439: It is recommended to use `self.log('val_mean_recall', ..., sync_dist=True)` when logging on epoch level in distributed setting to accumulate the metric across devices.
```

So here, `metrics` is the return value of calling `MetricCollection` with my logits and true masks, then I iterate over it like this: `for metric_name, metric in metrics.items()`, and finally properly log `metric`, not `metric.item()`.

I will test whether this still happens without `MetricCollection` a bit later.

**Comment by srprca** [comment]
This still happens when logged "properly" (without `.item()`) without `MetricCollection` wrapper.

**Comment by awaelchli** [comment]
Can you show it with a runnable example based on https://github.com/Lightning-AI/pytorch-lightning/blob/master/examples/pytorch/bug_report/bug_report_model.py?

**Comment by geometrikal** [comment]
> If you pass in scalar tensors then not of course. Then the warning is normal and expected. For logging TorchMetrics you would just pass in the metric directly into `self.log`. You can find a guide here: https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html

I get the same thing when logging using the manual method.

Going to try directly logging with the metric, but does that support ClasswiseWrapper?

**Comment by david-rohrschneider** [comment]
Hello, i can confirm the confusion. I am just training on 2 GPUs and cannot find any documentation on how to use MetricCollection in distributed environments. Im not using sync_dist, so getting the same warning and i am not sure if my metrics are computed / logged properly

**Comment by HenrikAsmuth** [comment]
Has there been any update on this issue recently? I searched the changelogs of lightning and torchmetrics but couldn't find any. For starters, similar to @david-rohrschneider, I would simply like to know if metrics are synced across devices or not.

**Comment by david-rohrschneider** [comment]
@HenrikAsmuth Hi, I stepped through the lightning code in debugger mode and can confirm they are distributed and computed correctly across devices, even though the warning is thrown. The message appears because I am not logging the Metric Object itself, but call the "compute" method before manually (because I need to remove a value there before logging). When calling compute, every TorchMetric has an internal distribute/reduce functionality. If you then log the resulting dict, torch lightnings log method does not know the dict comes from TorchMetrics, thus throwing the warning.

**Comment by HenrikAsmuth** [comment]
Thanks @david-rohrschneider . Was this now when running lightning without explicitly setting `sync_dist=True` ?

**Comment by david-rohrschneider** [comment]
@HenrikAsmuth exactly. Running sync_dist=True would actually mess this up again if you DONT pass the Metric Object itself and just the dict. If yiu pass the Metric object however, the argument is completely ignored (on purpose).

**Comment by SkafteNicki** [comment]
Nicki here, long time lightning contributor both on lightning side and torchmetrics side. I consider this page on the torchmetrics side of things to be best up-to-date regarding information on logging torchmetrics in lightning:
https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html
It gives example of the two recommended approaches:
* Logging the metric object directly
* Calling `forward`, `update` and `compute` manually and then logging using `self.log`

I think the TLDR is that torchmetrics has its own distributed sync logic to make sure the metrics are calculated correctly and if one is using these in lightning, they should ignore lightning metric sync logic. If one is unsure what to do, the clear recommendation is to log the metric object. In this case we can clearly detect what is happening and deal with it correctly. On the other hand if you log a tensor or dict of tensors, lightning has no idea if this value came from a torchmetrics object that can do the sync or this was a more "manually" calculated metric where you want to rely on the build-in sync of lightning.

**Comment by david-rohrschneider** [comment]
@SkafteNicki Thank you for your response! Could this be added in the documentation exactly like you explained? I think it would clarify the issue. However, in my case I was using MeanAveragePrecision and it produces a dict, where one entry is NOT a scalar tensor ("classes"). I also added it to a MetricCollection, further complicating things. It was not possible to directly log the metric object, since this raises the error of a non scalar tensor value. Therefore, I had to manually compute it which caused my confusion.

**Comment by SkafteNicki** [comment]
@david-rohrschneider that is a fair point. The main problem you are facing is not `MetricCollection` which is natively supported by `self.log_dict` similar to how a standard metric is natively support by `self.log`, the problem is (as you have identified) that `MeanAveragePrecision` outputs a dict and not a scalar-tensor which is not natively supported.

We can definitely figure out adding more to the documentation...I guess you would like this to be added to lightnings documentation?

**Comment by david-rohrschneider** [comment]
@SkafteNicki Got it, thanks! I saw the docs state that issue explicitly for MetricCollection (https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html#common-pitfalls). So this would probably be one sentence, explaining that the scalar-tensor behavior not only affects MetricCollection, but also Metrics like MAP.

However, I would prefer a solution where all Metric Classes (or at least the predefined ones) are forced to return scalar tensors so that they work out of the box. E.g. for MAP I am not sure if anyone uses the "classes" field and it could also be extracted manually via a different class property if someone needs it.
